### ISSUE TITLE
Errors in some interior node dates

### ISSUE THREAD

**Issue by jrosindell** [OneZoom/tree-build#18]
http://www.onezoom.org/life/@Ginglymostomatidae=331731 and the nearby Whale Shark (Rhincodon typus) all show parent nodes at 37.5 Kya which is quite suspect as in practice the data is probably missing - the next node up in the tree is 33 Mya which is much more sensible.

**Comment by hyanwong** [comment]
This is the relevant file:

https://github.com/OneZoom/tree-build/blob/main/data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Naylor2012Selachimorpha.PHY

It looks like I rescaled the dates using the PathD8 hack, so it wouldn't be surprising if there was an rounding error somewhere.

**Comment by davidebbo** [comment]
FWIW, here is what I see in `AllLife_full_tree.phy`, which uses `Naylor2012Selachimorpha.PHY`:

```
(
  (
    (
      (
        Stegostoma_fasciatum_ott122336
      )Stegostoma_ott122337
    )Stegostomatidae_ott696403,
    (
      Pseudoginglymostoma_brevicaudatum_ott24097
    )Pseudoginglymostoma_ott356286
  ):0.0,
  (
    (
      Rhincodon_typus_ott214509
    )Rhincodon_ott5264328
  )Rhincodontidae_ott738324
):33.28144
```

**Comment by davidebbo** [comment]
The behavior is the same in `tree-build`, so no difference between old and new repo.

I'm not familiar with this edge length / age logic, but I will try to debug it to see what it going on.

**Comment by davidebbo** [comment]
### TL;DR: how the blind legless lizard beat up the mighty whale shark

After some deep investigation, I've concluded that the issue is caused by a problem in a completely different part of the tree. This is due to the dual Post Order and then Pre Order logic we use in set_node_ages().

There are probably other problematic places, but let's focus on a specific one, [Dibamus bourreti](https://www.onezoom.org/life/@Dibamus_bourreti=413261). The branch length are coming from https://github.com/OneZoom/tree-build/blob/main/data/OZTreeBuild/AllLife/BespokeTree/include_noAutoOTT/Tetrapods_Zheng_base.PHY:

```
      (
        (
          Dibamus_bourreti:45.860264,
        ):18.298951,
      ):140.878302,
```

The problem is that these add up to 205.037517, when it really should be just 205 (which is the Squamata age).

The reason it's specifically affecting the whale shark (and some of his buddies) is that it is missing branch lengths, and so are its parent & grandparent (as seen in my comment above).

So what's happening is:
- On the Post Order (i.e. leaves first) date phase, the whale shark doesn't get a date at all, since it's missing branch length
- The extra 37517 years from the Dibamus bourreti branch ends up propagating all the way up the **entire tree**. e.g. `AllArchaeaAndEukaryotes_` ends up with an age of 3200.037518!
- Then on the Pre Order phase (parents first), we propagate the problems down to leaves that don't already have an age, including the whale shark. And we end up with a 'leftover' of 37517, because the sum of all the edges for the whale shark add up to exactly 3200. So we do 3200.037518 - 3200, and assign that to the whale shark / zebra shark / nurse shark concestor.

I would say there are two main ways to fix this:
1. The right way is to fix the edge lengths in Tetrapods_Zheng_base to add up to the correct number.
2. The tactical fix is to just set an explicit edge length of 0 for the whale shark (etc) in Naylor2012Selachimorpha.PHY, as @hyanwong suggests.

**Comment by hyanwong** [comment]
Ha - very well triaged. Thanks @davidebbo.

The problem is that we will always have some rounding error in the node times, so I'm not sure we can guarantee the tree will be exactly "ultrametric", as it is called.

I will have a think how to deal with this.

**Comment by davidebbo** [comment]
I think we could use a heuristic to prevent rounding issues from propagating too far.

Let's look at the entire chain for Dibamus bourreti, noting how our 37,518 year error makes its way all the way to the beginning of life:

| Taxon | Edge length | Age |
| ---------------- | --------- | - |
| Dibamus bourreti | 45.860264 | 0 |
| None | 18.298951 | 45.860264 |
| None | 140.878302 | 64.159216 |
| Squamata | 70.0 | 205.037518 |
| Lepidosauria | 25.0 | 275.037518 |
| Sauropsida | 20.0 | 300.037518 |
| Amniota | 20.0 | 320.037518 |
| Tetrapoda | 75.0 | 340.037518 |
| Dipnotetrapodomorpha | 5.0 | 415.037518 |
| Sarcopterygii | 10.0 | 420.037518 |
| Euteleostomi | 30.0 | 430.037518 |
| Gnathostomata | 65.0 | 460.037518 |
| Vertebrata | 10.0 | 525.037518 |
| ChordataMinusAmphioxus_ | 5.0 | 535.037518 |
| Chordata | 10.0 | 540.037518 |
| Deuterostomia | 10.0 | 550.037518 |
| Nephrozoa | 10.0 | 560.037518 |
| Bilateria | 20.0 | 570.037518 |
| BilatCnid_ | 10.0 | 590.037518 |
| BiCnCt_ | 20.0 | 600.037518 |
| BiCnCtPl_ | 30.0 | 620.037518 |
| Metazoa | 150.0 | 650.037518 |
| CollarClade_ | 180.0 | 800.037518 |
| Filozoa | 120.0 | 980.037518 |
| Holozoa | 250.0 | 1100.037518 |
| Opisthokonta | 350.0 | 1350.037518 |
| None | 100.0 | 1700.037518 |
| Obazoa | 150.0 | 1800.037518 |
| Amorphea | 50.0 | 1950.037518 |
| Podiata | 150.0 | 2000.037518 |
| Eukaryota | 550.0 | 2150.037518 |
| LE clade_ | 100.0 | 2700.037518 |
| TACKLE clade_ | 200.0 | 2800.037518 |
| ETACKLE clade_ | 200.0 | 3000.037518 |
| AllArchaeaAndEukaryotes_ | None | 3200.037518 |

The propagation happens because we always use the longest length from all the paths:

```python
                    if node.parent_node.age < (node.age + l):
                        node.parent_node.age = node.age + l
```

But we could tweak this by rounding `node.age + l` to eliminate non-significant digits. e.g. it could be something like saying that for anything older than 100 Ma, we round up to the closest 100,000 years. So in the case above, `Squamata` would stay at 205 Ma, and it wouldn't go any further. Or maybe a slightly more elaborate scheme along the same lines.

**Comment by davidebbo** [comment]
@hyanwong and @jrosindell I thought through various approaches. Some were a little too disrupting, so I settled on the following: during the 'percolate down' phase of set_node_ages(), never set an age under 50k years. The reasoning is that if such recent nodes were to legitimately exist, they should come from the initial 'up' phase (from the leaf). When they occur in the down phase, they are an unwanted effect of the issue described above (non-ultrametricity).

Here is the complete list of taxa that is affected by this change (coming from the debug output):

```
Not setting age of 0.03751800000009098 on Breviatea as it's too small (parent: Obazoa)
Not setting age of 0.03751799999997729 on Placozoa as it's too small (parent: BiCnCtPl_)
Not setting age of 0.0375179999999915 on None as it's too small (parent: Torpediniformes)
Not setting age of 0.03751799999997729 on SCYLIORHINIDAE3 as it's too small (parent: None)
Not setting age of 0.03751799999997729 on Most Carcharhinicae_ as it's too small (parent: None)
Not setting age of 0.03751799999997729 on SCYLIORHINIDAE2 as it's too small (parent: None)
Not setting age of 0.03751799999997729 on Ginglymostomatidae as it's too small (parent: None)
Not setting age of 0.03751799999997729 on None as it's too small (parent: None)
Not setting age of 0.0375179999999915 on Dalatiidae as it's too small (parent: None)
Not setting age of 0.03751799999997729 on SomniosidaeOxynotidae as it's too small (parent: None)
Not setting age of 0.03751799999997729 on Etmopteridae as it's too small (parent: None)
Not setting age of 0.03751799999997729 on Squatinidae as it's too small (parent: None)
Not setting age of 0.03751799999997729 on Pristiophoridae as it's too small (parent: None)
```

Every one of them is in fact affected by the 37,518 year issue, and I'm quite confident that they should all be zero age (which they are with the fix).

Please let me know what you think.

**Comment by hyanwong** [comment]
This sounds plausible.

If we have very recently extinct things, like the dodo and the solitaire, we represent them using an extra inserted null branch of the correct length, like this:

```
((:0.00025)Pezophaps_solitaria:10.75952,(:0.00035)Raphus_cucullatus:10.75942):6)
```

(We call this an "extinction prop"). This means that all tips *should* be at time 0, even if they are of extinct species.

**Comment by davidebbo** [comment]
Interesting, I had not encountered this "extinction prop" handling. Though indeed, that is not concerning for the proposed change.

**Comment by hyanwong** [comment]
After discussion, we decided that the best way would be to enforce ultrametricity in the input trees, and then figure out a way to stitch them together so that the resulting composite tree is also ultrametric.

At some point, the stitching algorithm is likely to require knowledge of "root branch" lengths (i.e. the length of the branch between the root of the input tree and the backbone. This can be adjusted as necessary to enforce ultrametricity of the output tree. At the moment there are 3 (!) potential sources for this branch length, which is confusing, especially in the case of conflicts. We decided that it would be a good idea to overhaul this. 

In fact, this could be a place where we specify an *absolute* date on the node. At the moment we do this:

`(((A:1,B:1)C:1,D@:2)E:1,F:3)` + `D = (X:0.5,Y:0.5):1.5` -> `(((A:1,B:1)D:1,(X:0.5,Y:0.5):1.5)E:1,F:3)`

Here we have 2 sources of the branch length from D to E:

1. the value from the first tree `D@:2` length which (since D is a single tip) should mark the absolute time before present, and
2. the value from the second (embedded) tree, given by `):1.5`, which is a relative length.

We had agreed that it would be good to be able to enforce absolute values for some branch points, and this seems a good opportunity to do so. I think at the moment I take the `2` value to be the length of the root branch, which I now think is the wrong approach.

I suggest that, if the root branch length is not calculable from ultrametricity constraints:
* if an absolute value (`D@:2`) exists, we calculate the required root branch length from that
* If there is no absolute value (e.g. we have `(((A:1,B:1)C:1,D@)E:1,F:3)`, then we take the relative root branch length from the embedded tree
* if there is no value in either, we raise an error

If there are conflicts (e.g. we can calculate the root branch from ultrametricity constraints but it conflicts with values from the trees) then we could either warn, or raise an error.

I feel that this would all require a bit of unit testing on simple example trees!

**Comment by davidebbo** [comment]
@hyanwong you cover two possible sources of branch length in the D@ scenario. Note that there is actually a third, which is in fact what the code (both old and new) favors. It is the length that comes from [substitution mapping](https://github.com/OneZoom/tree-build/blob/7dbc3b536730819f9408f89f1305e3e3f07d1dcc/oz_tree_build/tree_build/token_to_oz_tree_file_mapping.py) (which is a direct translation of the [old mapping](https://github.com/OneZoom/OZtree/blob/2d6a8487ff6133b473575e258ec3435169eec8fc/OZprivate/data/OZTreeBuild/AllLife/AdditionalFiles/base_tree.js) that you know better).

So the logic (both old and new, and I matched it) is:
1. Use the mapping length if available
2. Fall back to the value in the embedded tree (1.5 in your example)
3. We never fall back to the 2 in `D@:2` (not sure the history behind this logic)

Given this, I'm thinking that if we want to start working with absolute ages, it is better to just add a new field to our mapping file, rather than give atypical semantic to out Newick-ish file.

e.g. instead of:
```python
"PORIFERA": {"file": "PoriferaOneZoom.phy", "edge_length": 50, "taxon": None},
```

We could have:
```python
"PORIFERA": {"file": "PoriferaOneZoom.phy", "age": 600, "taxon": None},
```

**Comment by hyanwong** [comment]
Yes, you are right that there is a source in the substitution mapping. But I think it might be better to remove that parameter from the substitution API and just use the `D@:2` value (which should be equivalent to an absolute value, I think). 

If we do that, then each  tree (even with substitutions in it) should be ultrametric, I think. This might make it easier to check?

**Comment by davidebbo** [comment]
On a different note (unrelated to the @ included files), as I'm starting to dive into ultrametricity in the bespoke files, one initial observation is that there are two distinct types of rounding issues:
1. Sometimes, the edge lengths are completely correct in the bespoke file, but we end up with a very small rounding error at runtime when we add them up, due to FPU limitations. e.g. in Python, `1.1+0.1=1.2000000000000002`
2. In other cases, the lengths in the bespoke file have rounding errors baked into them. e.g. in GarsDeepfin.phy, for Atractosteus tristoechus, the edges (from leaf to file root) are `2.5+2.873563+16.241946+246.28449`, which add up to 267.899999. There is no 'rounding error' in this addition, but it's the **input** that is wrong.

My plan is to deal with this in the following way:
- For 1, no change is needed to the bespoke file since it is correct. Instead, we should perform rounding during processing when we add things up. e.g. `round(1.1+0.1, 10)==1.2`. This way, we don't propage these small errors too far.
- For 2, our best bet is to simply hand fix the bespoke files. e.g. in GarsDeepfin.phy, if I simply tweak `:16.241946` to `:16.241947`, the whole file becomes ultrametric. Given that we don't have that many files, it doesn't seem too painful to go through such a manual round of adjustments.

**Comment by davidebbo** [comment]
> If we do that, then each tree (even with substitutions in it) should be ultrametric, I think. This might make it easier to check?

Oh yes, that is a valid point. One way or another, if we can move away from having 3 distinct sources of truth, it's a step forward :)

**Comment by hyanwong** [comment]
> * For 1, no change is needed to the bespoke file since it is correct. Instead, we should perform rounding during processing when we add things up. e.g. `round(1.1+0.1, 10)==1.2`. This way, we don't propagate these small errors too far.
> * For 2, our best bet is to simply hand fix the bespoke files. e.g. in GarsDeepfin.phy, if I simply tweak `:16.241946` to `:16.241947`, the whole file becomes ultrametric. Given that we don't have that many files, it doesn't seem too painful to go through such a manual round of adjustments.

Yes, I completely agree with this.

**Comment by hyanwong** [comment]
> > If we do that, then each tree (even with substitutions in it) should be ultrametric, I think. This might make it easier to check?
> 
> Oh yes, that is a valid point. One way or another, if we can move away from having 3 distinct sources of truth, it's a step forward :)

Yes, I agree. The other thing that inclines me to remove the parameter to substitute is that, if we want to allow valid newick formats, then `D@:2` should be *allowable* (even if the `:2` is ignored). It would be easier to allow it and use it rather than silently ignore it, or bomb out with a newick-valid file.

**Comment by davidebbo** [comment]
> If we have very recently extinct things, like the dodo and the solitaire, we represent them using an extra inserted null branch of the correct length, like this:
> 
> ```
> ((:0.00025)Pezophaps_solitaria:10.75952,(:0.00035)Raphus_cucullatus:10.75942):6)
> ```
> 
> (We call this an "extinction prop"). This means that all tips _should_ be at time 0, even if they are of extinct species.

I want to make sure I correctly understand this. Here, when we check for ultrametricity, we want to include the extinction prop in the full path length, right? So in this case, we have 0.00025+10.75952+6 for Solitaire, and 0.00035+10.75942+6 for Dodo. They both equal 16.75977, which is a quirky number, but at least if makes the fragment ultrametric.

OTOH, looking at `(:0.0005)Dinornis_robustus:58.1`, I suspect that we really want this to be `(:0.0005)Dinornis_robustus:58.0995` instead, so that they add up to 58.1 (which makes sense in the context of the rest of the file). So I will make this change. Same goes for `((:0.0004)Aepyornis_maximus:17.2,(:0.0004)Mullerornis_agilis:17.2)Aepyornithidae:32.8`, where we'll want to subtract 0.0004 from both 17.2.

Do I have this right?

**Comment by hyanwong** [comment]
Yes, that is all correct.

**Comment by davidebbo** [comment]
I fixed a number of files in PR #35. What's interesting is that now the the 'big' 37.5 Kya issue is gone, we still get hit by a smaller instance of the same issue coming from elsewhere in the tree. e.g. now the Whale Shark parent has age 4.6 Kya. Basically, any instance anywhere in the tree where things add up to a little more than they should is going to cause this. And hopefully to can squash them all...

Now I'm struggling with PasserinesOneZoom.phy. The challenge is:
- Unlike most files, the age of the file root is not a sane number, but seems to be about 70.786162.
- It has 558 variants of that age, spread across ~6000 species. e.g. 70.7861619836, 70.78616198566, 70.78616199179, etc...
- The edge lengths are specified to 9 or 10 decimals, which seems unnecessary. Going to 6 decimals already gives us a one day granularity.
- It ends with `PASSERIFORMES:3.999641884;` which is completely different from the 8 coming from the mapping: `"PASSERIFORMES": {"file": "PasserinesOneZoom.phy", "edge_length": 8},`. The 8 is what we in fact use today, so the 3.999641884 is ignored.
- I think we can't really look at this file in a vacuum, as its age needs to also fit will in the context of its parent Neognathae_minus_passerines_OneZoom.PHY, and its own parent Tetrapods_Zheng_base.PHY, which all have numbers that are not well rounded.

I realize that a lot of this is a result of some PATHd8 magic. We should discuss if we can bring some sanity to this.

**Comment by davidebbo** [comment]
The tree is now fully ultrametric, and the original issue is addressed
